### PR TITLE
2023_qrack_raid_0 blog post

### PR DESCRIPTION
**Pull request by WrathfulSpatula**
This is the blog post about Qrack's frontier experiments in recreating something equivalent to 2019 Sycamore (or universal random circuit sampling in general) requested by Will.

(Merry Christmas, and Happy Holidays! This is low priority, but I stole a few minutes for it during our family's celebrations. See you in 2024!)

### CODE REVIEW

File: src/content/blog/2023_qrack_raid_0.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+---
+title: "How to build a ~40 qubit simulator for less than $3000 (and then simulate 54 qubits on it)!"
+author: Dan Strano
+day: 25
+month: 12
+year: 2023
```
Comment (nathanshammah) [inline src/content/blog/2023_qrack_raid_0.md:6]:
```suggestion
day: 5
month: 2
year: 2024
```

File: src/content/blog/2023_qrack_raid_0.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+IBM was likely the first to point out, after the 2019 Sycamore universal random circuit sampling experiments, that disk storage could be employed to increase effective RAM limits of quantum computer simulation far past what is attainable with GPU and “DIMMs” (RAM boards inserted into your computer motherboard). Roughly, the largest supercomputers in the world, even arrayed together, might only have DRAM capacity for about 49 or 50 qubits of state vector simulation, aside from other techniques like tensor networks or any approximate simulation techniques. (Just as with your personal computer hardware, adding 1 qubit _doubles_ the necessary memory footprint, and the capacity of the largest supercomputers in the world, today, does not double in negligible time, effort, or cost!) However, disk storage can be _nearly_ as good as “DIMMs,” but drastically cheaper and far easier to scale to terabyte-range and past!
```
Comment (nathanshammah) [inline src/content/blog/2023_qrack_raid_0.md:13]:
```suggestion
IBM was likely the first to [point out](https://research.ibm.com/blog/on-quantum-supremacy), after the 2019 Sycamore universal random circuit sampling experiments, that disk storage could be employed to increase effective RAM limits of quantum computer simulation far past what is attainable with GPU and “DIMMs” (RAM boards inserted into your computer motherboard). Roughly, the largest supercomputers in the world, even arrayed together, might only have DRAM capacity for about 49 or 50 qubits of state vector simulation, aside from other techniques like tensor networks or any approximate simulation techniques. (Just as with your personal computer hardware, adding 1 qubit _doubles_ the necessary memory footprint, and the capacity of the largest supercomputers in the world, today, does not double in negligible time, effort, or cost!) However, disk storage can be _nearly_ as good as “DIMMs,” but drastically cheaper and far easier to scale to terabyte-range and past!
```

File: src/content/blog/2023_qrack_raid_0.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+The obvious problem is that data transfer to and from disk is _far_ slower than “DIMMs.” Ultimately, “swap disk” as (random-access) memory needs to somehow be made fast enough to keep up with the processing throughput capacity of whatever CPU(s) to which it’s attached. However, Unitary Fund and the Qrack simulator development team have demonstrated proof-of-concept for creating swap disk resources that are _nearly_ fast enough to keep up with a 64-hyper-thread AMD Epyc processor: this prototype machine uses “RAID 0” configuration on 8 (“consumer-grade”) NVMe solid-state drives, achieving 8 times the throughput of a single drive! By “striping” the (single) logical swap disk partition across 8 hardware drives, interspersing small segments of memory in a pattern of “stripes” that are likely to be accessed at once across physical drives, the overall transfer speed nets to the _sum_ of the transfer speeds of each respective disk, achieving speeds close to sufficient to keep the CPU barely fully utilized on large state vector simulations (particularly with OS-level swap disk compression enabled, such as by manually enabling “zswap” with “zstd” compressor in Linux)!
```
Comment (nathanshammah) [inline src/content/blog/2023_qrack_raid_0.md:15]: Add link? Suggestions: arXiv report and or previous blog post on Qrack white paper.

File: src/content/blog/2023_qrack_raid_0.md (hunk 2 above)
Comment (nathanshammah) [reply src/content/blog/2023_qrack_raid_0.md:13]: Also, can you spell out once the exact words for the acronym DIMMs?

File: src/content/blog/2023_qrack_raid_0.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
+---
+title: "How to build a ~40 qubit simulator for less than $3000 (and then simulate 54 qubits on it)!"
+author: Dan Strano
+day: 25
+month: 12
+year: 2023
+---
+
+At 32-bit floating point precision (with 64-bit complex numbers), 32 *giga*bytes of memory will fit 32 qubits worth of state vector simulation. For each additional qubit desired, you need to _double_ the memory (compounded per additional qubit). 40 qubits, depending on precision, therefore requires 8 or 16 *tera*bytes (_256 or 512 times_ as much RAM). Unfortunately, typical ram DIMMs might come in 16 gigabyte increments (give-or-take up to a couple of factors of 2), and maybe you’ll fit 8 of them in a server motherboard, for 128 gigabytes. Alternatively, NVIDIA A100 GPUs might carry 80 gigabytes apiece, you might fit about 64 gigabytes worth of quantum state vector simulation in a single GPU, and then it will take 128 of these (with high-speed interconnects) to reach 8 terabytes, at a cost of thousands or tens of thousands of dollars per hour to rent. What is the hobbyist quantum computing researcher to do?
```
Comment (nathanshammah) [inline src/content/blog/2023_qrack_raid_0.md:9]: This post start is already very technical. Is there a way to make it a bit more gentle for non experts?

One suggestion would be to repeat the question of the title, with a bit more context. For example, something like: ```
There is a fast pace in the qubit count growth of experiments on quantum computers. At the same time, there is a growing literature of classical simulation (_add links_) of such experiments, often involving approximate methods and high performance computing (HPC) resources. One could thus be wondering: 

_What is the limit of what can be achieved on a personal computer/with moderate cloud resources?_

In this blog post we provide a possible recipe to this answer, showing how to build a ~40 qubit simulator for less than $3000. 
```

Still, even with this preamble, I feel like the current initial text starts a bit technical. Could you maybe provide a bit of context on how to do the resource estimation, without introducing immediately technical jargon? That could help a lot!

File: src/content/blog/2023_qrack_raid_0.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+When using the Qrack gate-based quantum computer simulator library, we officially suggest 64-bit “double” precision (with 128-bit complex numbers) when state vector simulation will exceed about 32 qubits: 16 TB of swap disk will realistically support a single state vector simulation up to 8 TB in width, giving you 39 qubits. However, we want to do something even more ambitious: we want our $2,500 machine to compete with 2019 Sycamore! In this case, we don’t need near-perfect fidelity of simulation, and we can approximate a circuit comparable to that Sycamore experiment with 32-bit “float” precision (with 64-bit complex numbers) and recourse to Qrack’s novel “hybrid” algorithms for simulation, including “quantum binary decision diagrams” (QBDD), based on the work of Robert Wille and team at Jülich Supercomputing Centre!
```
Comment (nathanshammah) [inline src/content/blog/2023_qrack_raid_0.md:29]: Please add a link to Qrack.

File: src/content/blog/2023_qrack_raid_0.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
...le the gate set used is not exactly that of the 2019 Sycamore experiment, this set is designed to (likely) be more expressive in less circuit depth while being easier to simulate, on average. It intersperses single-qubit gate layers with maximally densely packed 2-qubit “nearest-neighbor” couplers across all qubits on a (close to) “square chip” similar to the 2019 Sycamore experiment topology. Each single-qubit gate entails a random Pauli basis transformation across all Pauli bases, composed with a uniformly random variational “RZ” gate (Z-axis rotation), such that it could take as few as 3 layers of single-qubit gates to achieve a completely general 3-parameter unitary (“U”) gate. Each 2-qubit gate is a coupler on a nearest neighbor topology, with a quarter of all gate options being the 3 controlled Pauli gates, a quarter being 3 “anti-controlled” Pauli gates (being gated by “0” control state instead of “1” control state), and half of all gates being variants of the “swap” gate with 6 physically distinct choices of phase effects in addition to the “swap” component, with a 50/50 “coin-flip” chance of choosing either direction between control and target qubits for all 2-qubit gates.
```
Comment (nathanshammah) [inline src/content/blog/2023_qrack_raid_0.md:31]: please add relevant link.

**Review by nathanshammah — CHANGES_REQUESTED** [review]
Thanks @WrathfulSpatula and sorry this has taken so long. I update the date: can you please also update the title of the file to 2024? 

I made some initial comments. My feedback is a bit of summarizing/cutting text, if possible could help, together with some introductory subtitles along the way.

**Comment by KallieFerguson** [comment]
Happy to see there's movement here! If this gets published next week could you send a note to Frances so she can share on socials? 

I see it also wishes people happy holidays at the end, we might want to change that now that it's February!

**Comment by amirebrahimi** [comment]
@WrathfulSpatula, how come you decided not to post this? The draft was fun to read though. I appreciate the details of the build-out.

**Pull request closed without merging**